README
Welcome to snag, a package for (remote) monitoring under Nagios.

This package competes primarily with the NRPE mechanism used to perform
checks on remote hosts.  Chief advantages:

* It uses SSH as transport, so it's taking advantage of a package with
robust security, and which you probably have to have running anyway to
admin your servers.  You don't have to open up a separate service for
monitoring.

* (It can use other transports; ultimately check_snag just has to be able
to run something that causes snag to be run on the appropriate host with
standard output coming back.  Netcat, xinetd, or the like could be pressed
into service.  However, SSH is what was intended.)

* You get a free SSH service check out of this.

* You get a command you can run directly on the host as a quick
health-check (especially after making system changes!).

* It runs several checks per session, rather than requiring multiple
sessions to be opened per monitoring cycle.  Most of the standard checks
require a handful of syscalls and can be done more efficiently as a
batch.  

* Since it's enumerating mounted filesystems, it will do the set of checks
appropriate to the particular host.  If /var is a filesystem, you get a
/var check, otherwise not.

* You don't have to configure each check separately; a small Perl script
is included to check for check results coming back and propose Nagios
config file additions.

* You get a single check turning red if the host crashes or becomes
unreachable; it doesn't falsely claim that all the filesystems suddenly
filled up, etc, as the result of failed NRPE checks.

* You don't get several NRPE checks hitting simultaneously and causing a
load spike negatively affecting the host's primary function.  Snag is
single-threaded so its maximum load is limited.

How it works
------------

The snag executable generates output that looks something like this:

Disk space on /;1;DISK WARNING - / 2578 MB (19%) free space|/=2578MB
Disk inodes on /;0;INODES OK - / 1052605 inodes (59%) free|/=1052605 inodes
Disk space on /home;1;DISK WARNING - /home 9953 MB (19%) free space|/home=9953MB
Disk inodes on /home;0;INODES OK - /home 6458748 inodes (98%) free|/home=6458748 inodes
Load average;0;LOAD OK - load average: 0.57 0.55 0.40|load1=0.57 load5=0.55 load15=0.40
Uptime;0;UPTIME OK - up 21 days|uptime=1825881s
Swap free;0;SWAP OK - 99% of swap free|swap=99
Processes;0;PROCS OK - 148 total processes|procs=148

There is a config file that can be used to tune thresholds for these
standard checks, and to add Nagios Plugin checks.

The check_snag executable runs on the Nagios server, configured as a
normal Nagios active check.  It gets run by Nagios at the configured
schedule, and reports OK, WARNING, CRITICAL depending on what happened
when it tried to run snag on the target host.  Check results it got back
are fed to the Nagios external command pipe as passive check results.

Installation
------------

See the file INSTALL in the distribution.

Author/Copyright
----------------

Copyright (C) 2006 Anthony de Boer, released under GPL v2.



ChangeLog
2008-02-04 21:47:33 -0500  Anthony de Boer  <adb@adb.ca>

    * Version 0.5

2008-02-04 21:29:56 -0500  Anthony de Boer  <adb@adb.ca>

    * cvsignore/gitignore

2008-02-04 21:40:03 -0500  Anthony de Boer  <adb@adb.ca>

    * commenting

2008-02-04 21:10:28 -0500  Anthony de Boer  <adb@adb.ca>

    * cleanup

2008-02-04 21:06:21 -0500  Anthony de Boer  <adb@adb.ca>

    * statusword cleanup

2008-02-04 20:55:45 -0500  Anthony de Boer  <adb@adb.ca>

    * cleanup

2008-02-04 20:13:06 -0500  Anthony de Boer  <adb@adb.ca>

    * add .gitignore

2008-02-04 20:12:31 -0500  Anthony de Boer  <adb@adb.ca>

    * add list of potential checks to TODO

2008-01-16 11:07:48 -0500  Anthony de Boer  <adb@adb.ca>

    * doc tweaks

2008-01-16 11:06:54 -0500  Anthony de Boer  <adb@adb.ca>

    * skip swap check on swapless machines

2008-01-15 18:38:39 -0500  Anthony de Boer  <adb@adb.ca>

    * Merge branch 'master' into massage

2008-01-15 18:37:43 -0500  Anthony de Boer  <adb@adb.ca>

    * Merge commit '49e63bf1e50760fb7ff54d5a48ceef4aa45bd85f' into massage

2008-01-15 18:37:22 -0500  Anthony de Boer  <adb@adb.ca>

    * Merge commit '97aad9ae374360990faf0a7cc665978fe8bc5a99' into massage

2008-01-15 18:34:27 -0500  Anthony de Boer  <adb@adb.ca>

    * Merge commit 'a4b8bbaabacd36366c2394468a3509c9f912fb54' into massage

2008-01-14 22:58:14 -0500  Anthony de Boer  <adb@adb.ca>

    * Merge /home/adb/git/snag

2008-01-14 22:22:54 -0500  Anthony de Boer  <adb@adb.ca>

    * add 0.4 source tree

2008-01-14 22:19:20 -0500  Anthony de Boer  <adb@adb.ca>

    * add 0.3 source tree

2008-01-14 22:15:03 -0500  Anthony de Boer  <adb@adb.ca>

    * add 0.2 source tree

2006-03-18 22:34:21 +0000  adb  <adb>

    * release

2006-03-18 22:29:54 +0000  adb  <adb>

    * updated

2006-03-18 22:29:30 +0000  adb  <adb>

    * 8-second timeout

2006-03-18 22:28:44 +0000  adb  <adb>

    * work with huge filesystems

2006-03-18 21:45:59 +0000  adb  <adb>

    * updated

2006-03-04 19:27:17 +0000  adb  <adb>

    * new INSTALL file

2006-03-02 22:08:09 +0000  adb  <adb>

    * 0.3

2006-03-02 22:07:21 +0000  adb  <adb>

    * enhance

2006-03-02 21:18:15 +0000  adb  <adb>

    * new README

2006-02-25 14:31:19 +0000  adb  <adb>

    * GPL boilerplate

2006-02-25 14:27:34 +0000  adb  <adb>

    * new

2006-02-25 14:27:00 +0000  adb  <adb>

    * documented

2006-02-25 14:26:39 +0000  adb  <adb>

    * correction

2006-02-25 14:26:31 +0000  adb  <adb>

    * renamed smaster to check_snag

2006-02-25 14:26:09 +0000  adb  <adb>

    * renamed to check_snag.c

2006-02-25 14:25:51 +0000  adb  <adb>

    * new, was smaster.c

2006-02-25 13:35:56 +0000  adb  <adb>

    * dist target

2006-02-25 13:30:25 +0000  adb  <adb>

    * idea

2006-02-25 13:30:10 +0000  adb  <adb>

    * closeconfig

2006-02-25 13:24:26 +0000  adb  <adb>

    * various

2006-02-25 13:23:58 +0000  adb  <adb>

    * cleanup

2006-02-25 13:23:45 +0000  adb  <adb>

    * defaultdumper

2006-02-25 13:23:26 +0000  adb  <adb>

    * command processing

2006-02-25 13:22:55 +0000  adb  <adb>

    * new

2006-02-25 11:01:19 +0000  adb  <adb>

    * integrate parser

2006-02-25 11:00:25 +0000  adb  <adb>

    * update

2006-02-25 11:00:11 +0000  adb  <adb>

    * configfile

2006-02-25 11:00:02 +0000  adb  <adb>

    * expand

2006-02-25 10:59:45 +0000  adb  <adb>

    * remove testing code

2006-02-25 10:30:27 +0000  adb  <adb>

    * new

2006-02-18 17:32:26 +0000  adb  <adb>

    * updated

2005-12-28 19:49:44 +0000  adb  <adb>

    * initial


thresh.c
/* thresh.c - threshold processing
 *
 *	Copyright (C) 2006 Anthony de Boer
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of version 2 of the GNU General Public License as
 *	published by the Free Software Foundation.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with this program; if not, write to the Free Software
 *	Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307
 *	USA
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "snag.h"

#define HASHSIZE 1009

/*
 * threshold data is kept in a hash table so that we can look up each
 * discovered bit of information, see if there are thresholds, and set
 * severity appropriately.
 */

int hash(char *s) {
	int h = 0;
	while(*s) {
		h += *s++;
		}
	return (h % HASHSIZE);
	}

struct hnode {
	char *name;
	struct hnode *next;
	long lcrit, lwarn, hwarn, hcrit;
	};

struct hnode *htab[HASHSIZE];

int showitems = 0;

struct hnode *hasher(char *s, int make) {
	int h = hash(s);
	struct hnode *p = htab[h];
	char *s2;
	while(p) {
		if (strcmp(s, p->name) == 0) {
			return p;
			}
		p = p->next;
		}
	if (make && (s2 = strdup(s)) && (p = (struct hnode *)calloc(1, sizeof (struct hnode)))) {
		p->name = s2;
		p->next = htab[h];
		htab[h] = p;
		}
	return p;
	}

void hashadd(char *s, long lcrit, long lwarn, long hwarn, long hcrit) {
	struct hnode *hent = hasher(s, 1);
	if (!hent) {
		fprintf(stderr, "memory allocation failure\n");
		exit(1);
		}
	hent->lcrit = lcrit;
	hent->lwarn = lwarn;
	hent->hwarn = hwarn;
	hent->hcrit = hcrit;
	if (showitems) {
		printf("limit %s %ld %ld %ld %ld\n", s, lcrit, lwarn, hwarn, hcrit);
		}
	}

void hinit(int showdefaults) {
	showitems = showdefaults;
	hashadd("Disk_space", 3, 20, 110, 120);
	hashadd("Disk_inodes", 3, 20, 110, 120);
	hashadd("load1", -1, -1, 15, 30);
	hashadd("load5", -1, -1, 10, 25);
	hashadd("load15", -1, -1, 5, 20);
	hashadd("Uptime", -1, 3600, 42768000, 42949672);
	hashadd("Swap_free", 10, 50, 110, 120);
	hashadd("Processes", -1, -1, 150, 200);
	if (showdefaults) exit(0);
	}
	
int thresher(char *name, long val) {
	struct hnode *hent = hasher(name, 0);
	int rc;
	if (hent) {
		if (val > hent->lwarn && val < hent->hwarn) {
			rc = 0;
			}
		else if (val > hent->lcrit && val < hent->hcrit) {
			rc = 1;
			}
		else {
			rc = 2;
			}
		}
	else {
		rc = 3;
		}
	return rc;
	}

int gthresher(char *name, long val) {
	int c = thresher(name, val);
	if (c == 3) c = 4;
	return c;
	}

int gnormalize(int c) {
	if (c & 4) {
		c = 3;
		}
	else if (c & 2) {
		c = 2;
		}
	return c;
	}

